File: src/filters/rg.ts
import { cleanText } from "../formatting/ansi.js";
import { limitLines } from "../formatting/limits.js";
import type { FilterOptions, FilterResult } from "./types.js";

interface MatchLine {
  line: number;
  text: string;
}

function requestsStructuredOutput(command: readonly string[]): boolean {
  return command.some((part) => part === "--json" || part.startsWith("--json="));
}

export function filterRgOutput(input: string, options: FilterOptions): FilterResult {
  const cleaned = cleanText(input);
  const lines = cleaned.split("\n");

  if (requestsStructuredOutput(options.command ?? [])) {
    return {
      output: cleaned,
      kind: "rg",
      omitted: false,
      notes: ["explicit rg structured output preserved"],
    };
  }

  const byFile = new Map<string, MatchLine[]>();
  let parseable = 0;

  for (const line of lines) {
    const match = line.match(/^(.+?):(\d+):(.*)$/);
    if (!match) continue;
    const filePath = match[1] ?? "";
    const lineNumber = Number.parseInt(match[2] ?? "0", 10);
    const body = match[3] ?? "";
    if (!filePath || !Number.isFinite(lineNumber) || lineNumber <= 0) continue;

    const list = byFile.get(filePath) ?? [];
    list.push({ line: lineNumber, text: body.trim() });
    byFile.set(filePath, list);
    parseable += 1;
  }

  if (parseable === 0) {
    return {
      output: cleaned,
      kind: "rg",
      omitted: false,
      notes: ["unrecognized rg format; returned cleaned output"],
    };
  }

  const files = [...byFile.entries()].sort((left, right) => right[1].length - left[1].length);
  const output: string[] = [`matches: ${parseable} in ${files.length} files`];
  let omittedMatches = 0;

  for (const [filePath, matches] of files) {
    output.push(`${filePath} (${matches.length})`);
    const visible = matches.slice(0, Math.max(1, options.perFileLines));
    for (const item of visible) {
      output.push(`  ${item.line}: ${item.text}`);
    }
    if (matches.length > visible.length) {
      const hidden = matches.length - visible.length;
      omittedMatches += hidden;
      output.push(`  ... ${hidden} matches omitted ...`);
    }
  }

  const limited = limitLines(output, options.maxLines);
  return {
    output: limited.lines.join("\n"),
    kind: "rg",
    omitted: omittedMatches > 0 || limited.omitted > 0,
    notes: omittedMatches > 0 ? [`omitted ${omittedMatches} rg matches`] : [],
  };
}

File: internal/filters/rg.go
package filters

import (
	"regexp"
	"sort"
	"strconv"
	"strings"

	"github.com/ivanduplenskikh/shrinker/internal/formatting"
)

func applyRg(input string, options Options) Result {
	cleaned := formatting.CleanText(input)
	if hasStructuredOption(options.Command) {
		return Result{Output: cleaned, Kind: "rg"}
	}
	pattern := regexp.MustCompile(`^(.+?):(\d+):(.*)$`)
	type match struct {
		line int
		text string
	}
	byFile := map[string][]match{}
	parseable := 0
	for _, line := range strings.Split(cleaned, "\n") {
		parts := pattern.FindStringSubmatch(line)
		if len(parts) != 4 {
			continue
		}
		lineNumber, err := strconv.Atoi(parts[2])
		if err != nil || lineNumber <= 0 {
			continue
		}
		byFile[parts[1]] = append(byFile[parts[1]], match{lineNumber, strings.TrimSpace(parts[3])})
		parseable++
	}
	if parseable == 0 {
		return Result{Output: cleaned, Kind: "rg"}
	}
	files := make([]string, 0, len(byFile))
	for name := range byFile {
		files = append(files, name)
	}
	sort.Slice(files, func(i, j int) bool { return len(byFile[files[i]]) > len(byFile[files[j]]) })
	output := []string{"matches: " + strconv.Itoa(parseable) + " in " + strconv.Itoa(len(files)) + " files"}
	omitted := 0
	for _, name := range files {
		matches := byFile[name]
		output = append(output, name+" ("+strconv.Itoa(len(matches))+")")
		visible := min(len(matches), max(1, options.PerFileLines))
		for _, item := range matches[:visible] {
			output = append(output, "  "+strconv.Itoa(item.line)+": "+item.text)
		}
		if visible < len(matches) {
			hidden := len(matches) - visible
			omitted += hidden
			output = append(output, "  ... "+strconv.Itoa(hidden)+" matches omitted ...")
		}
	}
	limited, lineOmitted := formatting.LimitLines(output, options.MaxLines)
	return Result{Output: strings.Join(limited, "\n"), Kind: "rg", Omitted: omitted > 0 || lineOmitted > 0}
}
